from PyQt5 import QtCore, QtGui, QtWidgets
from PyQt5.QtWidgets import QMessageBox, QMainWindow, QStyledItemDelegate, QAction, QLineEdit
import sys

import machine_learning
from source import predict



class UI_linear_regression(QMainWindow):
    def __init__(self, parent=None):
        super(UI_linear_regression, self).__init__(parent)

    def setupUi(self):
        self.setObjectName("LinearRegression")
        self.resize(796, 540)
        self.centralwidget = QtWidgets.QWidget(self)
        self.setWindowIcon(QtGui.QIcon("../icon/logo.png"))
        self.centralwidget.setObjectName("centralwidget")
        self.columns = QtWidgets.QListWidget(self.centralwidget)
        self.columns.setGeometry(QtCore.QRect(20, 70, 301, 191))
        self.columns.setObjectName("columns")
        self.label_2 = QtWidgets.QLabel(self.centralwidget)
        self.label_2.setGeometry(QtCore.QRect(20, 280, 91, 21))
        font = QtGui.QFont()
        font.setFamily("Ubuntu Condensed")
        font.setPointSize(10)
        self.label_2.setFont(font)
        self.label_2.setObjectName("label_2")
        self.line_2 = QtWidgets.QFrame(self.centralwidget)
        self.line_2.setGeometry(QtCore.QRect(320, 70, 20, 421))
        self.line_2.setFrameShape(QtWidgets.QFrame.VLine)
        self.line_2.setFrameShadow(QtWidgets.QFrame.Sunken)
        self.line_2.setObjectName("line_2")
        self.label_5 = QtWidgets.QLabel(self.centralwidget)
        self.label_5.setGeometry(QtCore.QRect(350, 80, 141, 21))
        font = QtGui.QFont()
        font.setFamily("Ubuntu Condensed")
        font.setPointSize(10)
        self.label_5.setFont(font)
        self.label_5.setObjectName("label_5")
        self.label_6 = QtWidgets.QLabel(self.centralwidget)
        self.label_6.setGeometry(QtCore.QRect(350, 160, 121, 21))
        font = QtGui.QFont()
        font.setFamily("Ubuntu Condensed")
        font.setPointSize(10)
        self.label_6.setFont(font)
        self.label_6.setObjectName("label_6")
        self.intercept = QtWidgets.QLabel(self.centralwidget)
        self.intercept.setStyleSheet("color:red")

        self.intercept.setGeometry(QtCore.QRect(490, 160, 251, 21))
        font = QtGui.QFont()
        font.setFamily("Ubuntu Condensed")
        font.setPointSize(10)
        self.intercept.setFont(font)
        self.intercept.setObjectName("intercept")
        self.label_7 = QtWidgets.QLabel(self.centralwidget)
        self.label_7.setGeometry(QtCore.QRect(20, 320, 141, 21))
        font = QtGui.QFont()
        font.setFamily("Ubuntu Condensed")
        font.setPointSize(10)
        self.label_7.setFont(font)
        self.label_7.setObjectName("label_7")
        self.test_data = QtWidgets.QLineEdit(self.centralwidget)
        self.test_data.setGeometry(QtCore.QRect(160, 320, 81, 23))
        self.test_data.setObjectName("test_data")
        self.label_8 = QtWidgets.QLabel(self.centralwidget)
        self.label_8.setGeometry(QtCore.QRect(20, 360, 91, 21))
        font = QtGui.QFont()
        font.setFamily("Ubuntu Condensed")
        font.setPointSize(10)
        self.label_8.setFont(font)
        self.label_8.setObjectName("label_8")
        self.label_9 = QtWidgets.QLabel(self.centralwidget)
        self.label_9.setGeometry(QtCore.QRect(20, 400, 91, 21))
        font = QtGui.QFont()
        font.setFamily("Ubuntu Condensed")
        font.setPointSize(10)
        self.label_9.setFont(font)
        self.label_9.setObjectName("label_9")
        self.train_size = QtWidgets.QLabel(self.centralwidget)

        self.train_size.setGeometry(QtCore.QRect(110, 360, 181, 16))
        font = QtGui.QFont()
        font.setFamily("Ubuntu Condensed")
        font.setPointSize(10)
        self.train_size.setFont(font)
        self.train_size.setObjectName("train_size")
        self.test_size = QtWidgets.QLabel(self.centralwidget)
        self.test_size.setGeometry(QtCore.QRect(110, 400, 181, 16))
        font = QtGui.QFont()
        font.setFamily("Ubuntu Condensed")
        font.setPointSize(10)
        self.test_size.setFont(font)
        self.test_size.setObjectName("test_size")
        self.train = QtWidgets.QPushButton(self.centralwidget)
        self.train.setStyleSheet("background-color:rgb(94, 186, 125)")
        self.train.clicked.connect(self.train_liner)
        self.train.setGeometry(QtCore.QRect(70, 440, 141, 41))
        font = QtGui.QFont()
        font.setFamily("Ubuntu Condensed")
        font.setPointSize(10)
        self.train.setFont(font)
        self.train.setObjectName("train")
        self.output = QtWidgets.QPushButton(self.centralwidget)
        self.output.setGeometry(QtCore.QRect(350, 380, 111, 41))
        font = QtGui.QFont()
        font.setFamily("Ubuntu Condensed")
        font.setPointSize(10)
        self.output.setFont(font)
        self.output.setObjectName("output")
        self.line_3 = QtWidgets.QFrame(self.centralwidget)
        self.line_3.setGeometry(QtCore.QRect(330, 210, 451, 20))
        self.line_3.setFrameShape(QtWidgets.QFrame.HLine)
        self.line_3.setFrameShadow(QtWidgets.QFrame.Sunken)
        self.line_3.setObjectName("line_3")
        self.label_11 = QtWidgets.QLabel(self.centralwidget)
        self.label_11.setGeometry(QtCore.QRect(350, 260, 201, 21))
        font = QtGui.QFont()
        font.setFamily("Ubuntu Condensed")
        font.setPointSize(10)
        self.label_11.setFont(font)
        self.label_11.setObjectName("label_11")
        self.label_12 = QtWidgets.QLabel(self.centralwidget)
        self.label_12.setGeometry(QtCore.QRect(350, 290, 201, 21))
        font = QtGui.QFont()
        font.setFamily("Ubuntu Condensed")
        font.setPointSize(10)
        self.label_12.setFont(font)
        self.label_12.setObjectName("label_12")
        self.label_13 = QtWidgets.QLabel(self.centralwidget)
        self.label_13.setGeometry(QtCore.QRect(350, 320, 201, 21))
        font = QtGui.QFont()
        font.setFamily("Ubuntu Condensed")
        font.setPointSize(10)
        self.label_13.setFont(font)
        self.label_13.setObjectName("label_13")
        self.mae = QtWidgets.QLabel(self.centralwidget)
        self.mae.setStyleSheet("color:red")
        self.mae.setGeometry(QtCore.QRect(540, 260, 251, 16))
        font = QtGui.QFont()
        font.setFamily("Ubuntu Condensed")
        font.setPointSize(10)
        self.mae.setFont(font)
        self.mae.setObjectName("mae")
        self.mse = QtWidgets.QLabel(self.centralwidget)
        self.mse.setStyleSheet("color:red")
        self.mse.setGeometry(QtCore.QRect(540, 290, 251, 16))
        font = QtGui.QFont()
        font.setFamily("Ubuntu Condensed")
        font.setPointSize(10)
        self.mse.setFont(font)
        self.mse.setObjectName("mse")
        self.rmse = QtWidgets.QLabel(self.centralwidget)
        self.rmse.setStyleSheet("color:red")
        self.rmse.setGeometry(QtCore.QRect(540, 320, 251, 16))
        font = QtGui.QFont()
        font.setFamily("Ubuntu Condensed")
        font.setPointSize(10)
        self.rmse.setFont(font)
        self.rmse.setObjectName("rmse")
        self.test_size_btn = QtWidgets.QPushButton(self.centralwidget)
        self.test_size_btn.clicked.connect(self.set_size)
        self.test_size_btn.setStyleSheet("background-color:rgb(94, 186, 125)")
        self.test_size_btn.setGeometry(QtCore.QRect(250, 320, 71, 23))
        font = QtGui.QFont()
        font.setFamily("Ubuntu Condensed")
        font.setPointSize(10)
        self.test_size_btn.setFont(font)
        self.test_size_btn.setObjectName("test_size_btn")
        self.weights = QtWidgets.QTextBrowser(self.centralwidget)
        self.weights.setGeometry(QtCore.QRect(490, 50, 291, 81))
        self.weights.setObjectName("weights")
        self.label_14 = QtWidgets.QLabel(self.centralwidget)
        self.label_14.setGeometry(QtCore.QRect(210, 0, 321, 51))
        font = QtGui.QFont()
        font.setFamily("Segoe UI Light")
        font.setPointSize(22)
        font.setBold(False)
        font.setItalic(False)
        font.setWeight(50)
        self.label_14.setFont(font)
        self.label_14.setStyleSheet("color: rgb(11, 34, 57);")
        self.label_14.setObjectName("label_14")
        self.target_2 = QtWidgets.QLabel(self.centralwidget)
        self.target_2.setGeometry(QtCore.QRect(110, 280, 181, 21))
        font = QtGui.QFont()
        font.setFamily("Ubuntu Condensed")
        font.setPointSize(10)
        self.target_2.setFont(font)
        self.target_2.setObjectName("target_2")
        self.predict = QtWidgets.QPushButton(self.centralwidget)
        self.predict.clicked.connect(self.predict_open)
        self.predict.setStyleSheet("background-color:rgb(94, 186, 125)")
        self.predict.setGeometry(QtCore.QRect(540, 440, 111, 41))
        font = QtGui.QFont()
        font.setFamily("Ubuntu Condensed")
        font.setPointSize(10)
        self.predict.setFont(font)
        self.predict.setObjectName("predict")
        self.label = QtWidgets.QLabel(self.centralwidget)
        self.label.setGeometry(QtCore.QRect(30, 50, 55, 16))
        self.label.setObjectName("label")
        self.number_col = QtWidgets.QLabel(self.centralwidget)
        self.number_col.setGeometry(QtCore.QRect(100, 50, 91, 16))
        self.number_col.setObjectName("number_col")
        self.label_link = QtWidgets.QLabel(self.centralwidget)
        self.label_link.hide()
        self.label_link.setGeometry(QtCore.QRect(580, 10, 55, 16))
        self.label_link.setObjectName("label_link")
        self.label_15 = QtWidgets.QLabel(self.centralwidget)
        self.label_15.setGeometry(QtCore.QRect(350, 350, 201, 21))
        font = QtGui.QFont()
        font.setFamily("Ubuntu Condensed")
        font.setPointSize(10)
        self.label_15.setFont(font)
        self.label_15.setObjectName("label_15")
        self.r2 = QtWidgets.QLabel(self.centralwidget)
        self.r2.setGeometry(QtCore.QRect(540, 350, 251, 16))
        self.r2.setStyleSheet("color:red")
        font = QtGui.QFont()
        font.setFamily("Ubuntu Condensed")
        font.setPointSize(10)
        self.r2.setFont(font)
        self.r2.setObjectName("r2")
        self.setCentralWidget(self.centralwidget)
        self.menubar = QtWidgets.QMenuBar(self)
        self.menubar.setGeometry(QtCore.QRect(0, 0, 796, 25))
        self.menubar.setObjectName("menubar")
        self.setMenuBar(self.menubar)
        self.statusbar = QtWidgets.QStatusBar(self)
        self.statusbar.setObjectName("statusbar")
        self.setStatusBar(self.statusbar)

        self.retranslateUi()
        # QtCore.QMetaObject.connectSlotsByName(LinearRegression)

    def retranslateUi(self):
        _translate = QtCore.QCoreApplication.translate
        self.setWindowTitle(_translate("LinearRegression", "Linear Regression"))
        self.label_2.setText(_translate("LinearRegression", "Target : "))
        self.label_5.setText(_translate("LinearRegression", "Coef (Weights) : "))
        self.label_6.setText(_translate("LinearRegression", "Intercept : "))
        self.intercept.setText(_translate("LinearRegression", ""))
        self.label_7.setText(_translate("LinearRegression", "Test Data Size:"))
        self.test_data.setText(_translate("LinearRegression", "0.1"))
        self.label_8.setText(_translate("LinearRegression", "Train Size:"))
        self.label_9.setText(_translate("LinearRegression", "Test Size:"))
        self.train_size.setText(_translate("LinearRegression", ""))
        self.test_size.setText(_translate("LinearRegression", ""))
        self.train.setText(_translate("LinearRegression", "Train"))
        self.output.setText(_translate("LinearRegression", "Show Graph"))
        self.label_11.setText(_translate("LinearRegression", "Mean Absolute Error:"))
        self.label_12.setText(_translate("LinearRegression", "Mean Square Error:"))
        self.label_13.setText(_translate("LinearRegression", "Root Mean Sq. Error:"))
        self.mae.setText(_translate("LinearRegression", ""))
        self.mse.setText(_translate("LinearRegression", ""))
        self.rmse.setText(_translate("LinearRegression", ""))
        self.test_size_btn.setText(_translate("LinearRegression", "Set"))
        self.label_14.setText(_translate("LinearRegression", "Linear Regression"))
        self.target_2.setText(_translate("LinearRegression", "<html><head/><body><p>Dropout or Graduate<br/></p></body></html>"))
        self.predict.setText(_translate("LinearRegression", "Predict"))
        self.label.setText(_translate("LinearRegression", "Columns:"))
        self.number_col.setText(_translate("LinearRegression", "TextLabel"))
        self.label_link.setText(_translate("LinearRegression", "TextLabel"))
        self.label_15.setText(_translate("LinearRegression", "R squared:"))
        self.r2.setText(_translate("LinearRegression", ""))

    def set_size(self):
        try:
            path = self.label_link.text()
            number = self.test_data.text()
            size = machine_learning.size_split(path, float(number))
            self.train_size.setText(str(size[0]))
            self.test_size.setText(str(size[1]))
        except:
            QMessageBox.information(self, 'System',
                                    'An error occurred while setting test, train size. Please try again!',
                                    QMessageBox.Close)

    def train_liner(self):
        try:
            number = self.test_data.text()
            path = self.label_link.text()
            result=machine_learning.liner(path,'error',float(number))
            #print(result)
            coef = ' '.join(map(str,result[0]))
            self.weights.setText(coef)
            #self.weights.setText(str(result[0]))
            self.intercept.setText(str(result[1]))
            self.mae.setText(str(result[2]))
            self.mse.setText(str(result[3]))
            self.rmse.setText(str(result[4]))
            self.r2.setText(str(result[5]))
        except:
            QMessageBox.information(self, 'System', 'An error occurred while training. Please try again!',
                                    QMessageBox.Close)
    def predict_open(self):
        try:
            self.ui = QMainWindow()
            self.ui = predict.Ui_predict_window()
            self.ui.setupUi()
            self.ui.label_link.setText(self.label_link.text())
            self.ui.label_tt.setText('Linear Regression')
            self.ui.show()
        except:
            QMessageBox.information(self, 'System', 'An error occurred while opening predict. Please try again!',
                                    QMessageBox.Close)
if __name__ == "__main__":
    app = QtWidgets.QApplication(sys.argv)
    ui = UI_linear_regression()
    ui.setupUi()
    ui.show()
    sys.exit(app.exec_())
